### DISCUSSION TITLE
[Schema] `ForeignKeyColumnDef` and `KeyColumnDef` column types are misleading.

### DISCUSSION THREAD

**Discussion by neumannjan** [LukasZahradnik/deep-db-learning#14] [General]
I initially meant the Schema to be used like this:

* If you want to know the primary key, you look at the `key` attribute of each column (NOT at whether a column has a `KeyColumnDef` type!)
* If you want to know the foreign keys, you look at the `foreign_keys` attribute (NOT at whether a column has a `ForeignKeyColumnDef` type!)
* If you want to know the type of the column purely in the sense of how the ML algorithm should see the column, then you look at the `ColumnDef` types
    * `KeyColumnDef` was supposed to mean "this column has no special meaning, other than it being part of a primary key"
        * however, this means that you can have e.g. a `NumericColumnDef` which is the primary key if you want to have the value itself as a feature! You will know that it is a primary key by looking at its `key` attribute.
    * `ForeignKeyColumnDef` was supposed to mean "this column has no special meaning, other than it being part of a foreign key"
        * however, this means that you can have e.g. a `NumericColumnDef` which is the foreign key if you want to have the value itself as a feature! You will know that it is a foreign key by looking at the `foreign_keys` attribute.

This means that if your approach to determining the primary key is "check all columns for the KeyColumnDef type", then you will not get correct results! Analogically for the foreign key as well.

However, clearly the `KeyColumnDef` and `ForeignKeyColumnDef` types are confusing (see #5 and #6), so maybe it is better to remove those two types and instead replace them with `OmitColumnDef`, to make it less confusing?

That way it would clearly state that "this column may or may not be a primary/foreign key, but with respect to ML features, it is not learned on as a feature per se".

(I intended `KeyColumnDef` and `ForeignKeyColumnDef` to do exactly that as well (i.e. a special case of `OmitColumnDef` pretty much), but clearly the purpose of the column types is misleading.)

**Comment by LukasZahradnik** [comment]
I found out that the PTE dataset (from relational fit) has the target column set as a foreign key, which slightly complicates things - the information, whether it is a categorical or numeric value, is missing.

**Reply by neumannjan** [reply]
Well, if it's set to either NumericColumnDef or to CategoricalColumnDef, then it automatically means that it shouldn't be missing.

**Reply by neumannjan** [reply]
Or does the Schema autodetect set it to ForeignKeyColumnDef? If yes, we could simply add the target column as a parameter to the detector, and for such column it will always try to guess a (feature-ful) type

**Reply by LukasZahradnik** [reply]
Yes, it's autodetected as ForeignKeyColumnDef. Having there a concrete type should be enough

**Reply by neumannjan** [reply]
Will fix tomorrow evening or so

**Reply by neumannjan** [reply]
Fixed in 260ba15
